pyflow/RCDT2/Nodes/core.py

```diff
@@ -0,0 +1,127 @@
+from typing import Literal
+from PyFlow.Core import NodeBase, PinBase
+import rclpy
+from rclpy.node import Node
+from threading import Thread
+from logging import getLogger
+from inflection import underscore
+
+from sensor_msgs.msg import Image
+
+logger = getLogger(__name__)
+
+
+class PinManager:
+    def __init__(self):
+        self.input_names: list[str] = []
+        self.input_pins: list[PinBase] = []
+
+        self.output_names: list[str] = []
+        self.output_pins: list[PinBase] = []
+
+    def add_input_pin(self, name: str, pin: PinBase) -> None:
+        self.input_names.append(name)
+        self.input_pins.append(pin)
+
+    def add_ouput_pin(self, name: str, pin: PinBase) -> None:
+        self.output_names.append(name)
+        self.output_pins.append(pin)
+
+    def get_pin_names(self, pin_type: Literal["input", "output"]) -> list[str]:
+        return getattr(self, f"{pin_type}_names")
+
+
+class PyflowRosBridge:
+    rclpy.init()
+    node = Node("pyflow")
+    Thread(target=rclpy.spin, args=[node], daemon=True).start()
+    node.get_logger().info("Node started!")
+
+
+class PyflowBase(NodeBase):
+    def __init__(self, name: str):
+        super().__init__(name)
+
+        self.pin_manager = PinManager()
+
+    def get_data(self, pin_name: str) -> object:
+        self.pin_manager.get_pin(pin_name)
+
+
+class PyflowComputer(PyflowBase):
+    def __init__(self, name: str):
+        super().__init__(name)
+
+    def compute(self, *_args: any, **_kwargs: any) -> None:
+        raise NotImplementedError
+
+
+class PyflowExecutor(PyflowBase):
+    def __init__(self, name: str):
+        super().__init__(name)
+
+    def execute(self, *_args: any, **_kwargs: any) -> None:
+        raise NotImplementedError
+
+
+class Service(PyflowExecutor):
+    def __init__(self, name: str):
+        super().__init__(name)
+        self.service: type
+
+        self.createInputPin("In", "ExecPin", callback=self.execute)
+        self.createOutputPin("Out", "ExecPin")
+        self.create_data_pins()
+
+        topic_name = underscore(self.service.__name__)
+        self.client = PyflowRosBridge.node.create_client(self.service, topic_name)
+
+    def execute(self, *_args: any, **_kwargs: any) -> None:
+        # Simple test on data passing:
+        data: Image = self.pin_manager.input_pins[0].getData()
+        logger.info(f"input = {data.header.frame_id}")
+        data = Image()
+        data.header.frame_id = "TEST"
+        self.pin_manager.output_pins[0].setData(data)
+
+    def create_data_pins(self) -> None:
+        service_parts = ["Request", "Response"]
```

Comment:
ipv hier zelf beide parts te benoemen kan dit beter in het aanroepen van deze methode gedaan worden (in de __init__). 

In de __init__ bepalen we voor werlke parts we data pins willen, en vervolgens roepen we deze methode aan met het juiste parts. Dan kan de buitenste for-loop ook uit deze methode gehaald worden.